flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteStateTable.java

```diff
@@ -219,985 +43,41 @@
-	 */
-	void releaseSnapshot(CopyOnWriteStateTableSnapshot<K, N, S> snapshotToRelease) {
-
-		Preconditions.checkArgument(snapshotToRelease.isOwner(this),
-				"Cannot release snapshot which is owned by a different state table.");
-
-		releaseSnapshot(snapshotToRelease.getSnapshotVersion());
-	}
-
-	// StateTableEntry -------------------------------------------------------------------------------------------------
-
-	/**
-	 * One entry in the {@link CopyOnWriteStateTable}. This is a triplet of key, namespace, and state. Thereby, key and
-	 * namespace together serve as a composite key for the state. This class also contains some management meta data for
-	 * copy-on-write, a pointer to link other {@link StateTableEntry}s to a list, and cached hash code.
-	 *
-	 * @param <K> type of key.
-	 * @param <N> type of namespace.
-	 * @param <S> type of state.
-	 */
-	@VisibleForTesting
-	protected static class StateTableEntry<K, N, S> implements StateEntry<K, N, S> {
-
-		/**
-		 * The key. Assumed to be immutable and not null.
-		 */
-		@Nonnull
-		final K key;
-
-		/**
-		 * The namespace. Assumed to be immutable and not null.
-		 */
-		@Nonnull
-		final N namespace;
-
-		/**
-		 * The state. This is not final to allow exchanging the object for copy-on-write. Can be null.
-		 */
-		@Nullable
-		S state;
-
-		/**
-		 * Link to another {@link StateTableEntry}. This is used to resolve collisions in the
-		 * {@link CopyOnWriteStateTable} through chaining.
-		 */
-		@Nullable
-		StateTableEntry<K, N, S> next;
-
-		/**
-		 * The version of this {@link StateTableEntry}. This is meta data for copy-on-write of the table structure.
-		 */
-		int entryVersion;
-
-		/**
-		 * The version of the state object in this entry. This is meta data for copy-on-write of the state object itself.
-		 */
-		int stateVersion;
-
-		/**
-		 * The computed secondary hash for the composite of key and namespace.
-		 */
-		final int hash;
-
-		StateTableEntry(StateTableEntry<K, N, S> other, int entryVersion) {
-			this(other.key, other.namespace, other.state, other.hash, other.next, entryVersion, other.stateVersion);
-		}
-
-		StateTableEntry(
-			@Nonnull K key,
-			@Nonnull N namespace,
-			@Nullable S state,
-			int hash,
-			@Nullable StateTableEntry<K, N, S> next,
-			int entryVersion,
-			int stateVersion) {
-			this.key = key;
-			this.namespace = namespace;
-			this.hash = hash;
-			this.next = next;
-			this.entryVersion = entryVersion;
-			this.state = state;
-			this.stateVersion = stateVersion;
-		}
-
-		public final void setState(@Nullable S value, int mapVersion) {
-			// naturally, we can update the state version every time we replace the old state with a different object
-			if (value != state) {
-				this.state = value;
-				this.stateVersion = mapVersion;
+	CopyOnWriteStateMapSnapshot<K, N, S>[] getStateMapSnapshotArray() {
+		CopyOnWriteStateMapSnapshot<K, N, S>[] snapshotArray =
+			new CopyOnWriteStateMapSnapshot[state.length];
```

Comment:
I would suggest to suppress the warning for unchecked assignment here or, as this place is not really performance critical, rather just use array list instead of array.